src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+        throw std::logic_error("Failed to write positions table to ledger");
+      }
+
+      // Write positions table offset at start of file
+      if (fseeko(file, 0, SEEK_SET) != 0)
+      {
+        throw std::logic_error("Failed to set file offset to 0");
+      }
+
+      if (fwrite(&table_offset, sizeof(table_offset), 1, file) != 1)
+      {
+        throw std::logic_error("Failed to write positions table to ledger");
+      }
+
+      if (fflush(file) != 0)
+      {
+        throw std::logic_error(fmt::format("Failed to flush ledger file"));
+      }
+
+      completed = true;
+    }
+
+    bool compact(size_t idx)
+    {
+      if (!completed || compacted || (idx != get_last_idx()))
+      {
+        // No effect if compaction idx is not last idx
+        return false;
+      }
+
+      if (fflush(file) != 0)
+      {
+        throw std::logic_error(fmt::format("Failed to flush ledger file"));
+      }
+
+      fs::rename(
+        fs::path(dir) / fs::path(get_file_name()),
+        fs::path(dir) /
+          fs::path(fmt::format(
+            "{}-{}{}",
+            get_file_name(),
+            get_last_idx(),
+            ledger_committed_suffix)));
+
+      compacted = true;
+      return true;
+    }
+  };
+
+  class MultipleLedger
+  {
+  private:
+    ringbuffer::WriterPtr to_enclave;
+
+    // Ledger directory
+    const std::string ledger_dir;
+
+    // Keep tracks of all ledger files for writing.
+    // Current ledger file is always the last one
+    std::list<std::shared_ptr<LedgerFile>> files;
+
+    // Cache of ledger files for reading
+    size_t max_read_cache_size;
+    std::list<std::shared_ptr<LedgerFile>> files_read_cache;
+
+    const size_t chunk_threshold;
+    size_t last_idx = 0;
+    size_t compacted_idx = 0;
+
+    // True if a new file should be created when writing an entry
+    bool require_new_file;
+
+    auto get_it_contains_idx(size_t idx) const
+    {
+      if (idx == 0)
+      {
+        return files.end();
+      }
+
+      auto f = std::upper_bound(
+        files.begin(),
+        files.end(),
+        idx,
+        [](size_t idx, const std::shared_ptr<LedgerFile>& f) {
+          return (idx <= f->get_last_idx());
+        });
+
+      return f;
+    }
+
+    std::shared_ptr<LedgerFile> get_file_from_cache(size_t idx)
+    {
+      if (idx == 0)
+      {
+        return nullptr;
+      }
+
+      // First, try to find file from read cache
+      for (auto const& f : files_read_cache)
+      {
+        if (f->get_start_idx() <= idx && idx <= f->get_last_idx())
+        {
+          return f;
+        }
+      }
+
+      // If the file is not in the cache, find the file from the ledger
+      // directory
+      std::map<size_t, std::string> all_files;
+      for (auto const& f : fs::directory_iterator(ledger_dir))
+      {
+        all_files.emplace(
+          get_start_idx_from_file_name(f.path().filename()),
+          f.path().filename());
+      }
+
+      auto f_ = std::upper_bound(
```

Comment:
Do we need to build this map of all files? This map will include our unfinished write files, so there's a risk we add them to our read cache (looks like that's currently avoided because the only calling point checks the write-files first, but seems the intent of this function could be narrowed to ensure it only talks about finished, readable files). While iterating through the committed files, we can use the first one whose range includes our index?

Reply:
You're right. We can simply loop through all the files and only consider the ones that are committed (as you rightly pointed, the non-committed ones will be in the write cache). I've implemented this.